[WIP] Complete rework

NewEditOfficial opened the pull request:
**Description** 

This pull request focus on revamping/improving the backend logic overall, and adjusting the frontend to work perfectly with those changes.

State: WIP

**TODO** 

- [x] Create function which checks if user is able to perform those tasks: CREATE, PUBLISH, DELETE 
- [x] Create buy news paper function with proper validation.
- [x] On resource restart hook the clear of the ped from having any attached entities and so on.
- [X] Define support for ESX & ox_inventory.

NewEditOfficial committed:
feat: (new_version)

NewEditOfficial committed:
LOCALES

feat: (LOCALES): Added locales and basic translation function to it.

NewEditOfficial committed:
Delete futte-newspaper-master.zip

NewEditOfficial committed:
FRONTEND_DATA

chore: (FRONTEND): Adjusted logic for frontend with fetchStories, which allows to refresh data from backend.

NewEditOfficial committed:
BUILDED_ASSETS

BUILDED_ASSETS

NewEditOfficial committed:
DB

fix: (DATABASE): Added validation for each function

NewEditOfficial committed:
SYNC

fix: (RESOURCE_STARTUP): This should fix the sync issue, sometimes data not occuring on client.

NewEditOfficial committed:
NUI

feat: (NUI) Added resync  of data and handleMenu function which allows to show/hide menu.

NewEditOfficial committed:
STORIES_CACHE

feat: (STORIES): Create cache when insert data into database and sync them to all users on server.

NewEditOfficial committed:
TWEAKS

feat: (LOCALES): Added missing locales.
feat: (RESET_PED_RESOURCE_RESTART): Added reset state when resource is restarted.
fix: (IS_REPORTER_VALIDATION): Added function to validate if user has job as reporter to allow him perform specific actions.
chore: (STORIES_SORTING): Add sorting of recent stories by date, to show in proper order.

NewEditOfficial committed:
ASSETS

feat: (QBCORE): Define item list for qb-core based inventories
feat: (OX): Define item list for ox_inventory
chore: (INVENTORY_IMAGES): Moved to assets/inventoryIcons/

NewEditOfficial committed:
FRAMEWORKS

feat: (FRAMEWORK_API): Added basic framework bridge for ESX/QBCore
feat: (INVENTORY_API): Added basic hook of ox_inventory / qb-inventories that can be used with the resource.
feat: (INIT_DEBUG): Show on resource start relevant server/resource information.
chore: (CONFIG): Added supported inventories, some variables, refactor

NewEditOfficial committed:
RCORE_PRISON - API

chore: (DEBUG) Improve print to be look better + added future todo for frontend timestamp / date.

xFutte commented on server/framework/qbcore.lua:
```diff
@@ -0,0 +1,136 @@
+                    end
+                end
+
+                -- ?? Revamp to work with our job structure, so we can access all data from one place.
```
Missing or comment just never deleted?

NewEditOfficial replied:
Left over, will remove it.

xFutte commented on server/framework/standalone.lua:
```diff
@@ -0,0 +1,9 @@
+Framework.getPlayer = function(playerId)
+    if not playerId then return nil end
+
+    return {
+
+    }
```
Shouldn't this return the player and not an empty table?

NewEditOfficial replied:
I should define some some placeholder data, will see.

xFutte commented on server/sv-init.lua:
```diff
@@ -0,0 +1,34 @@
+                local jailTime = prisonerData and prisonerData.jailTime or 0
+                local prisonerName = prisonerData and prisonerData.name
+
+                -- TODO: Requires to modify frontend with some variable with prefix jail format (seconds/mins/hours/days) to make proper sense.
```
Something you need from me? `ps-core` is going to come with utility functions for formatting dates, so we can use that if that was the intention?

NewEditOfficial replied:
The thing is now that is strictly to months in frontend when you see, related to create definition for locales of UI? 🤔

xFutte commented on server/sv-main.lua:
```diff
@@ -0,0 +1,191 @@
+
+    if player and data then
+        if not IsReporter(playerId) then
+            Framework.showNotify(playerId, l('NOT_ALLOWED'), 'error')
+            return
+        end
+
+        local storyType = data.type and data.type:upper()
+
+        if storyType then
+            if STORIES_CACHE[storyType] then
+                local id = #STORIES_CACHE[storyType] + 1
+
+                STORIES_CACHE[storyType][id] = {
+                    title = data.title,
+                    story_type = storyType,
+                    image = data.image,
+                    jailed_player = '',
+                    jailed_time = 0,
+                    id = id,
+                    publisher = player.charName,
+                    body = data.body,
+                    date = data.date,
+                }
+
+                table.sort(STORIES_CACHE[storyType], function(a, b)
+                    return a.date > b.date
+                end)
+
+                Database.PublishStory(data, player.charName)
+
+                SyncStories(-1, STORIES_CACHE)
+
+                Framework.showNotify(playerId, l('STORY_PUBLISHED'), 'success')
+            else
+                print('futte-newspaper:server:publishStory: storyType is not valid')
+            end
+        else
+            print('futte-newspaper:server:publishStory: storyType is nil')
+        end
```
Some of it already has it, but could a couple of things be refactored to it has guardclauses instead? I think that would be more clean.

NewEditOfficial replied:
Sure we can improve this part.

xFutte commented on server/sv-main.lua:
```diff
@@ -0,0 +1,191 @@
+RegisterNetEvent('futte-newspaper:server:deleteStory', function(data)
+    local playerId = source
+    local player = Framework.getPlayer(playerId)
+    local storyId = data and data.id
```
I might be a bit nit picky, but this variable name is not so representative of what it does.

NewEditOfficial replied:
Will come with better variable naming.

xFutte commented on server/sv-main.lua:
```diff
@@ -0,0 +1,191 @@
+    return false
+end
+
+-- Framework.RegisterUsableItem(Config.Items and Config.Items.NEWSPAPER or 'newspaper', function(source, item)
+--     local playerId = source
+--     local itemCount = Framework.GetItemCount(playerId, item.name)
+
+--     if itemCount > 0 then
+--         TriggerClientEvent('futte-newspaper:client:openNewspaper', playerId)
+--     end
+-- end)
```
Is this somewhere else?

NewEditOfficial replied:
I think i have it somewhere else - left over.

xFutte commented on shared/init.lua:
```diff
@@ -0,0 +1,115 @@
+        local itemState = Framework.DoesItemExist(Config.Items.NEWSPAPER) or 'UNK'
+        local jobState = Framework.DoesJobExist(Config.Job.name) or 'UNK'
+        
+        sprint('\n^3%s^0\nVersion: ^3%s^0\nFramework: ^3%s^0\n^0Inventory: ^3%s\n^0\n^0Is required item [%s] registered: ^3%s^0\n^0Is required job [%s] registered in your FW: ^3%s\n^0', GetCurrentResourceName(), currentVersion, Config.Framework, Config.Inventory, Config.Items.NEWSPAPER, itemState and 'YES' or 'PLEASE DEFINE THE REQUIRED ITEM IN YOUR INVENTORY!', Config.Job.name, jobState and 'YES ...
```
Should this be `print` instead?

NewEditOfficial replied:
I will refactor this and define proper naming. 👍

xFutte commented on web/src/App.svelte:
```diff
@@ -7,37 +7,48 @@
+			const placeholderStory = {
+				id: 0,
+				title: 'Welcome to futte-newspaper',
+				body: "<p>futte-newspaper is a standalone ressource for FiveM. It has the following dependencies:</p><p><ul><li>qb-target</li><li>oxmysql</li></ul></p><p>I hope you'll enjoy the resource. Feel free to open issues if you find a bug/wish new functionality. This story will automatically delete itself as soon as you write a new one.</p><p>- xFutte</p>",
```
Well, I guess it is not a standalone resource..

NewEditOfficial replied:
Lets keep it ESX/QBCORE, standalone version could be sorted out in future if needed really. 👍

xFutte reviewed (changes requested):
A couple of comments. I'll re-review when updated/answered.

NewEditOfficial commented on server/framework/qbcore.lua:
```diff
@@ -0,0 +1,136 @@
+-- ?? Structure for framework jobs in general.
+
+-- [script:futte-newspap] {
+-- [script:futte-newspap]     "job": {
+-- [script:futte-newspap]         "name": "police",
+-- [script:futte-newspap]         "label": "LSPD",
+-- [script:futte-newspap]         "grade_label": "Lieutenant",
+-- [script:futte-newspap]         "grade": 3,
+-- [script:futte-newspap]         "grade_name": "lieutenant",
+-- [script:futte-newspap]         "onDuty": true
+-- [script:futte-newspap]     },
+-- [script:futte-newspap]     "charName": "Xdddd Adsasdada",
+-- [script:futte-newspap]     "charId": "char1:8a5e30da582cda4e9cd8d0a929bd3c852d9d8d23"
+-- [script:futte-newspap] }
```
Left over, will remove it.

NewEditOfficial commented on fxmanifest.lua:
```diff
@@ -1,19 +1,42 @@
 
-ui_page 'web/public/index.html'
-
+--[[ Resource Information ]]
```
Sure, what should be there? Let me know 👍

NewEditOfficial committed:
Update qbcore.lua

chore: (JOB_STRUCTURE_NOTE): Removed leftover note for myself when i was working with framework job data.

NewEditOfficial committed:
Update qbcore.lua

chore: (getPlayer): Note for framework job data, removed it.

NewEditOfficial committed:
Delete standalone.lua
